pkg/search/postgres/store_cache.go

```diff
@@ -312,7 +328,27 @@ func (c *cachedStore[T, PT]) Walk(ctx context.Context, fn func(obj PT) error) er
 
 // GetByQuery returns the objects from the store matching the query.
 func (c *cachedStore[T, PT]) GetByQuery(ctx context.Context, query *v1.Query) ([]*T, error) {
-	return c.underlyingStore.GetByQuery(ctx, query)
+	defer c.setCacheOperationDurationTime(time.Now(), ops.GetByQuery)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
```

Comment:
I think the cache lock can be acquired once the underlying store has returned the IDs to fetch.